File: code/C++/ABC177D.cpp
#include<iostream>
#include<vector>

class UnionFind {
    public:
        std::vector<int> par, siz;
        UnionFind(int n): par(n, -1), siz(n, 1){}

        //経路圧縮あり
        int root(int x){
            if(par[x] == -1) return x;
            return par[x] = root(par[x]);
        }

        //経路圧縮なし
        /*int root(int x){
            if(par[x] == -1) return x;
            return root(par[x]);
        } */

        bool unite(int x, int y){
            int rx = root(x), ry = root(y);
            if(rx == ry) return false;
            if(size(rx) < size(ry)) std::swap(rx, ry);
            
            par[ry] = rx;
            siz[rx] += siz[ry];
            return true; 
        }

        bool same(int x, int y){
            return root(x) == root(y);
        }

        int size(int x){
            return siz[root(x)];
        }

};

//解答の方針
//友達が一人もいないグループになるようにするには
//それぞれの人において（友達の数+自分の数）になることが必要です。
//したがって、友達が最も多い人の(友達の数 + 1)が答えになります

int main(){
    int n, m;
    std::cin >> n >> m;
    UnionFind uf(n);
    for(int i = 0;i < m;++i){
        int a, b;
        std::cin >> a >> b;
        --a;--b;
        uf.unite(a, b);
    }

    int ans = 1;
    for(int i = 0;i < n;++i){
        ans = std::max(ans, uf.size(i));
    }
    std::cout << ans << std::endl;

    return 0;
}

File: code/Python/ABC177D.py
class UnionFind:
    def __init__(self, n):
        self.par = [-1]*n
        self.siz = [1]*n

    #経路圧縮あり
    def root(self, x):
        if self.par[x] == -1:
            return x

        self.par[x] = self.root(self.par[x])
        return self.par[x]

    #経路圧縮なし
    """     
    def root(self, x):
        if self.par[x] == -1:
            return x
        return self.root(self.par[x])
    """
    
    def unite(self, x, y):
        rx = self.root(x)
        ry = self.root(y)
        if rx == ry:
            return False
        if self.size(rx) < self.size(ry):
            rx, ry = ry, rx

        self.par[ry] = rx
        self.siz[rx] += self.siz[ry]
        return True


    def same(self, x, y):
        return self.root(x) == self.root(y)

    def size(self, x):
        return self.siz[self.root(x)]

#解答の方針
#友達が一人もいないグループになるようにするには
#それぞれの人において（友達の数+自分の数）になることが必要です。
#したがって、友達が最も多い人の(友達の数 + 1)が答えになります

n, m = map(int, input().split())
uf = UnionFind(n)

for _ in range(m):
    a, b = map(int, input().split())
    a -= 1
    b -= 1
    uf.unite(a, b)

ans = 1
for i in range(n):
    ans = max(ans, uf.size(i))

print(ans)